### ISSUE TITLE
error occurred:multiple definition of `mlpack::rl::ContinuousActionEnv::Action conflicts.When the header file is referenced by multiple files, the reinforcement learning library conflicts

### ISSUE THREAD

**Issue by tongqimxd** [mlpack/mlpack#2854]
I'm new to mlpack. I want to do some simple experiments with reinforcement learning, but there are some problems.
I found that when the header file is referenced individually, it can be linked normally, but when multiple files are referenced, the links will conflict, as follows:
```
Scanning dependencies of target test
[ 33%] Building CXX object CMakeFiles/test.dir/main.cpp.o
[ 66%] Building CXX object CMakeFiles/test.dir/my_dqn.cpp.o
[100%] Linking CXX executable test
CMakeFiles/test.dir/my_dqn.cpp.o:(.bss+0x0): multiple definition of `mlpack::rl::ContinuousActionEnv::Action::size'
CMakeFiles/test.dir/main.cpp.o:(.bss+0x0): first defined here
CMakeFiles/test.dir/my_dqn.cpp.o:(.bss+0x8): multiple definition of `mlpack::rl::ContinuousActionEnv::State::dimension'
CMakeFiles/test.dir/main.cpp.o:(.bss+0x8): first defined here
CMakeFiles/test.dir/my_dqn.cpp.o:(.bss+0x10): multiple definition of `mlpack::rl::DiscreteActionEnv::Action::size'
CMakeFiles/test.dir/main.cpp.o:(.bss+0x10): first defined here
CMakeFiles/test.dir/my_dqn.cpp.o:(.bss+0x18): multiple definition of `mlpack::rl::DiscreteActionEnv::State::dimension'
CMakeFiles/test.dir/main.cpp.o:(.bss+0x18): first defined here
collect2: error: ld returned 1 exit status
CMakeFiles/test.dir/build.make:117: recipe for target 'test' failed
make[2]: *** [test] Error 1
CMakeFiles/Makefile2:94: recipe for target 'CMakeFiles/test.dir/all' failed
make[1]: *** [CMakeFiles/test.dir/all] Error 2
Makefile:102: recipe for target 'all' failed
make: *** [all] Error 2
```
The project is like this：
CMakeLists.txt
```
cmake_minimum_required(VERSION 3.7.0)
project(test)
SET(CMAKE_CXX_FLAGS "-std=c++11 -O3 -fopenmp")
set(CMAKE_INCLUDE_CURRENT_DIR ON)
add_compile_options(-fext-numeric-literals)
include_directories(mlpack)
add_executable(test main.cpp my_dqn.cpp)
target_link_libraries(test mlpack)
```
main.cpp
```
#include "my_dqn.h"

int main(int argc, char *argv[])
{
    return 0;
}
```
my_dqn.h
`#include <mlpack/methods/reinforcement_learning/environment/env_type.hpp>`
my_dqn.cpp
`#include <my_dqn.h>`
I don't know how to correct it.Thank you for your help~

**Comment by zoq** [comment]
Hello, to me it looks like if you put include guards (https://en.wikipedia.org/wiki/Include_guard) around implementation in  `my_dqn.h` in should work. To me it looks like `main.cpp` as well as `my_dqn.cpp` inlude `my_dq.h`.

**Comment by tongqimxd** [comment]
> Hello, to me it looks like if you put include guards (https://en.wikipedia.org/wiki/Include_guard) around implementation in `my_dqn.h` in should work. To me it looks like `main.cpp` as well as `my_dqn.cpp` inlude `my_dq.h`.

Thank you for your help. The same error still appears after modifying in the following way.
my_dqn.h
```
#ifndef GRAND_MY_DQN_H
#define GRAND_MY_DQN_H
#include <mlpack/methods/reinforcement_learning/environment/env_type.hpp>
#endif /* GRAND_MY_DQN_H */
```
I think because my_ dqn. h contains only one header file(env_type.hpp), which is put include guards.So no matter this file  my_dqn.h is protected or not, the result is the same.Thank you very much, but I'm still confused.

**Comment by jonpsy** [comment]
I think your problem stems from

```add_executable(test main.cpp my_dqn.cpp) ```

During the linking stage, both these files have the definition for your function. Which causes the linker to panic :D.

https://stackoverflow.com/questions/17904643/error-with-multiple-definitions-of-function

**Comment by tongqimxd** [comment]
> I think your problem stems from
> 
> `add_executable(test main.cpp my_dqn.cpp) `
> 
> During the linking stage, both these files have the definition for your function. Which causes the linker to panic :D.
> 
> **EDIT:**
> 
> The main problem is having definitions inside the header file. Include guard protect against including the same header twice for a single compilation unit, but it doesn't help in this [situation](https://stackoverflow.com/a/9576187).
> 
> I'd suggest
> a) Put contents of my_dqn.cpp inside main.cpp
> OR
> b) Put only the declarations of functions inside `my_dqn.h` and include `#include<mlpack/methods..../env_type.hpp>` in either of my_dqn.cpp or main.cpp

Thank you very much for your answer. I know where my question is. It's in< mlpack / methods... / env_ type.hpp >In the header file, there is a definition statement, which leads to an error when linking after multiple references. Thank you again.

**Comment by rcurtin** [comment]
Ah, I think we need to reopen this.  This actually is a bug in the mlpack implementation.

We are defining the variables `ContinuousActionEnv::State::dimension` and `ContinuousActionEnv::Action::size` as `static`, and then setting a value in the header file.  This means that these variables will occur in *both* `libmlpack.so` as well as any application linked against `libmlpack.so`, causing the multiple definition error.

I've opened a fix in #2863.  Thanks for the report! :+1:

**Comment by zoq** [comment]
With #2863 merged we can keep this one closed :+1:
